vowpalwabbit/guard.h

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+#pragma once
+
+#include <utility>
+#include "future_compat.h"
+
+// While it is a general concept this specific implemenation of scope based cleanup was inspired by
+// https://github.com/microsoft/wil/blob/cd51fa9d3c24e9e9f5d4cdf1a9768d68f441fc48/include/wil/resource.h#L543
+
+namespace VW
+{
+namespace details
+{
+template <typename T>
+class swap_guard_impl
+{
+ public:
+  swap_guard_impl(T& original_location, T& value_to_swap) noexcept : _original_location(original_location), _value_to_swap(value_to_swap)
+  {
+    std::swap(_original_location, _value_to_swap);
+  }
+
+  swap_guard_impl(const swap_guard_impl&) = delete;
+  swap_guard_impl& operator=(const swap_guard_impl&) = delete;
+
+  swap_guard_impl(swap_guard_impl&& other) noexcept
```

Comment:
the move constructors aren't necessarily safe for all of our objects. If they're not necessary, I'd rather delete them to be safe

Reply:
These are definitely needed as they permit the make functions.